### PR TITLE
Add LoongArch64 sha1 assembly support

### PR DESCRIPTION
**Pull request by heiher**
Add sha1 assembly support for [LoongArch64](https://doc.rust-lang.org/rustc/platform-support/loongarch-linux.html) (A tier-2 with host tools target).

### CODE REVIEW

File: sha1/src/lib.rs [hunk 1]
```diff
@@ -9,7 +9,12 @@
 //! [`sha-1`]: https://crates.io/crates/sha-1
 
 #![no_std]
-#[cfg(not(any(target_arch = "x86_64", target_arch = "x86", target_arch = "aarch64")))]
+#[cfg(not(any(
+    target_arch = "x86_64",
+    target_arch = "x86",
+    target_arch = "aarch64",
+    target_arch = "loongarch64"
+)))]
```
Comment (newpavlov) [inline sha1/src/lib.rs:17]: Add mention of LoongArh64 in the `compile_error!` message after this config.

**Review by newpavlov — COMMENTED** [review]
The assembly looks fairly straightforward.Can you measure whether it results in an improved performance when compared to the software backend in `sha1`?

**Comment by heiher** [comment]
> The assembly looks fairly straightforward.Can you measure whether it results in an improved performance when compared to the software backend in `sha1`?

Performance improved slightly. :smiling_face_with_tear: 

On Loongson 3A5000 @ 2.3GHz:

Without this patch:

```
test sha1_10    ... bench:          33 ns/iter (+/- 0) = 303 MB/s
test sha1_100   ... bench:         308 ns/iter (+/- 0) = 324 MB/s
test sha1_1000  ... bench:       2,559 ns/iter (+/- 2) = 390 MB/s
test sha1_10000 ... bench:      25,057 ns/iter (+/- 11) = 399 MB/s
```

With this patch:

```
test sha1_10    ... bench:          32 ns/iter (+/- 0) = 312 MB/s
test sha1_100   ... bench:         291 ns/iter (+/- 0) = 343 MB/s
test sha1_1000  ... bench:       2,450 ns/iter (+/- 1) = 408 MB/s
test sha1_10000 ... bench:      24,079 ns/iter (+/- 12) = 415 MB/s
```

File: sha1/src/lib.rs (hunk 1 above)
Comment (heiher) [reply sha1/src/lib.rs:17]: ~Oops! Why I can't see the latest changes after force push.~

**Comment by newpavlov** [comment]
Hm, the performance difference is relatively small, but well above noise levels, so I guess we can merge this PR. You could squeeze a bit more performance by implementing the looping over blocks inside assembly. It will allow you to keep hasher state in registers without spilling it into memory after processing each block (UPD: unfortunately, the spilling is necessary since we can not preserve old state value in registers).

Note that the asm-hashes repository crates are  effectively deprecated and we do not plan to use them in the next release cycle of the hashes repository crates. If you have time, it would be better to adapt this PR to using `asm!` directly inside the `sha1` crate.

**Comment by heiher** [comment]
Completed in PR: https://github.com/RustCrypto/hashes/pull/504

**Pull request closed without merging**